### PR TITLE
Bluetooth: controller: ISO-AL CIS RX framed and ISO bugfixes

### PR DESCRIPTION
**Pull request by mtpr-ot**
Adds reassembly of framed PDUs to ISO adaptation layer.
Reference times are also computed based on anchor points,
transmission latency components and a PDU specific time offset.

Some partial preparations for ISO broadcast can be found for
convenience. This needs to be properly merged, and currently will not
work with both CONFIG_BT_CTLR_CONN_ISO and CONFIG_BT_CTLR_SYNC_ISO
enabled.

Includes fixes and improvements for unframed mode.

Signed-off-by: Asger Munk Nielsen <asmk@oticon.com>
Signed-off-by: Morten Priess <mtpr@oticon.com>

### CODE REVIEW

**Comment by mtpr-ot** [comment]
The test `broadcast_iso.sh` fails. I assume this is due to some overlapping implementation. @cvinayak, could you please take a look at that?

**Review by cvinayak — CHANGES_REQUESTED** [review]
Overlapping PR: https://github.com/zephyrproject-rtos/zephyr/pull/40602

Please have a separate PR for Broadcast ISO related file changes in this PR, so that I can take independent ISOAL changes for my testing. We can integrate broadcast ISO changes together thereafter.

File: subsys/bluetooth/controller/Kconfig [hunk 1]
```diff
@@ -591,6 +591,22 @@ config BT_CTLR_BROADCAST_ISO
...
 	default BT_CTLR_ADV_ISO || BT_CTLR_SYNC_ISO
 
+config BT_CTLR_BROADCAST_ISO_STREAMS
+	int "LE Broadcast Isochronous Streams"
+	depends on BT_CTLR_BROADCAST_ISO
+	range 1 32
+	default 2
+	help
+	  Maximum supported total number of BISes.
+
```
[the remark this answers is not in the document]
Comment (mtpr-ot) [reply subsys/bluetooth/controller/Kconfig:601]: I agree that having the ISO-AL and broadcast in separate PRs makes perfect sense. However, It is quite difficult to separate BIS/BIG out, as the code has now been designed to incorporate both, and logic just looks strange when removing one. Will try and see if it's possible.

[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig]: Thank you. There is only a little conflict with my other PR now.

File: subsys/bluetooth/controller/hci/hci.c [hunk 2]
```diff
@@ -53,6 +53,8 @@
 #include "ll_sw/ull_sync_internal.h"
 #include "ll_sw/ull_conn_types.h"
 #include "ll_sw/ull_conn_internal.h"
+#include "ll_sw/isoal.h"
+#include "ll_sw/ull_iso_types.h"
```
Comment (cvinayak) [inline subsys/bluetooth/controller/hci/hci.c:57]: I have conflict here. Minor, nothing that cannot be resolved.

File: subsys/bluetooth/controller/ll_sw/ull_iso_types.h [hunk 3]
```diff
@@ -4,6 +4,33 @@
...
+#else
+#define IS_CIS_HANDLE(_handle) 0
+#endif /* CONFIG_BT_CTLR_CONN_ISO */
+
+/* Common memebers for ll_conn_iso_stream and ll_broadcast_iso_stream */
+struct ll_iso_stream_hdr {
+	struct ll_iso_datapath *datapath_in;
+	struct ll_iso_datapath *datapath_out;
+};
+
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_iso_types.h:33]: This one we can refactor after this PR is merged.

**Review by cvinayak — COMMENTED** [review]
Quick-ack, LGTM. Let me get this merged first, if its ok with you? https://github.com/zephyrproject-rtos/zephyr/pull/40602

**Review by cvinayak — APPROVED** [review]

**Review by thoh-ot — APPROVED** [review]

**Comment by mtpr-ot** [comment]
Rebased to include #40602. `ll_setup_iso_path` will need some refactoring to unite CIS- and BIS implementation. This PR will allow CIS **or** BIS, but not both configured at the same time.

**Comment by mtpr-ot** [comment]
> Rebased to include #40602. `ll_setup_iso_path` will need some refactoring to unite CIS- and BIS implementation. This PR will allow CIS **or** BIS, but not both configured at the same time.

@cvinayak, iso_broadcast test fails - can you please take a look? Thanks.

**Comment by cvinayak** [comment]
> @cvinayak, iso_broadcast test fails - can you please take a look? Thanks.

Sure. Working on it today.

**Comment by mtpr-ot** [comment]
Looks good, @cvinayak. Thanks

**Pull request merged**